File: R/npst.r
## ################################################################################
## Program : NPST
## Language: R
## This code is the wrapper function to call the shared library
## "npstfortran". The final result derives the empirical
## distribution of a generalization of Hewitt's and Rogerson's
## nonparametric seasonality tests.
##
## Copyright (C) 2004-2010  Roland Rau
## 
## This program is free software; you can redistribute it and/or
## modify it under the terms of the GNU General Public License
## as published by the Free Software Foundation; either version 2
## of the License, or (at your option) any later version.

## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.

## You should have received a copy of the GNU General Public License
## along with this program; if not, write to the Free Software
## Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
## ################################################################################

.First.lib <- function(lib, pkg)
  library.dynam("npst", pkg, lib)
     

npst <- function(long=12, peak=6, repts=100000, outputrows=10) {
  

  # what is the maximum rank sum?
  maxranksum <- sum(long:(long-peak+1))
  
  # initialize vector where results are stored with zeros.
  resvec <- rep(0, maxranksum)

  # initialize base data and repeat the values.
  basedata <- 1:long
  basedata2 <- rep(1:long,2)

  # calling external Fortran (90/95) routine to do the heavy work.
  outsourcedloop <- .Fortran("npst", PACKAGE="npst",
                             lng=as.integer(long),
                             pk=as.integer(peak),
                             rpts=as.integer(repts),
                             mxrksum=as.integer(maxranksum),
                             resvec=as.integer(resvec),
                             basedata=as.integer(basedata),
                             basedata2=as.integer(basedata2))

  # Calculate the p-values, i.e. the cumulative distribution function.
  pvals <- cumsum(outsourcedloop$resvec[order(maxranksum:1)])/repts
  maxisums <- maxranksum:1
  returndf <- data.frame(Max.Rank.Sum=maxisums, p.values=pvals)
  return(returndf[1:outputrows,])
}

File: src/npst.f90
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
! Subroutine: NPST
! Language:   Fortran 90/95
! This code can be compiled to a shared library (DLL on Windows) and
! then be interfaced via R (www.r-project.org) to derive the empirical
! distribution of a generalization of Hewitt's and Rogerson's
! nonparametric seasonality tests.
!
! This code has been successfully compiled  on Windows XP, GNU/Linux (Ubuntu) and FreeBSD
! using 'gfortran' from the GNU Compiler Collection (gcc)
!
! Copyright (C) 2007-2010  Roland Rau
!
! This program is free software; you can redistribute it and/or
! modify it under the terms of the GNU General Public License
! as published by the Free Software Foundation; either version 2
! of the License, or (at your option) any later version.

! This program is distributed in the hope that it will be useful,
! but WITHOUT ANY WARRANTY; without even the implied warranty of
! MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
! GNU General Public License for more details.

! You should have received a copy of the GNU General Public License
! along with this program; if not, write to the Free Software
! Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

subroutine npst(lng, pk, rpts, mxrksum, resvec, basedata, basedata2)
  implicit none
  integer               :: lng, pk, rpts, mxrksum
  integer               :: resvec(mxrksum)
  real                  :: x
  integer               :: basedata(lng)
  integer               :: basedata2(lng*2)
  integer               :: tempvalue, posis, maxconsecsum, temp

  integer               :: i,j

  integer               :: ii, n, clock

  integer, DIMENSION(:), ALLOCATABLE :: seed
  
  CALL RANDOM_SEED(size = n)
  ALLOCATE(seed(n))
  
  CALL SYSTEM_CLOCK(COUNT=clock)
  
  seed = clock + 37 * (/ (i - 1, i = 1, n) /)
  CALL RANDOM_SEED(PUT = seed)
  
  DEALLOCATE(seed)

  i = 0
 
  ! the main loop
  do j=1,rpts,1
     call random_number(x)
  
     ! generating a random permutation of 'basedata' and 'basedata2'
     ! 
     do i = lng, 1, -1
        posis = floor(1 + x*i)
        tempvalue = basedata(posis)
        basedata(posis)=basedata(i)
        basedata(i) = tempvalue
        basedata2(i) = basedata(i)
        basedata2(i+lng) = basedata(i)
     end do
     
    
     maxconsecsum = sum(basedata2(1:(1+pk-1)))
     temp =  sum(basedata2(1:(1+pk-1)))
      do i =2, lng, 1
         temp = temp - basedata2(i-1) + basedata2(i+pk-1)
         if (temp > maxconsecsum) maxconsecsum = temp
      end do
     
      resvec(maxconsecsum) = resvec(maxconsecsum) + 1

   end do
   !  deallocate(iseed)
end subroutine npst
